### PR TITLE
feat: enhance error handling in git-push and repo services

### LINKED ISSUE #1188: Improve UI error handling and API error codes
**Is your feature request related to a problem? Please describe.**
Right now, the UI handles errors rather inflexibly. Usually, we check for a single error code, and then send out a generic error message based on the code.

This is an example from `/src/ui/services/git-push.js`:

```js
const rejectPush = async (id, setMessage, setUserAllowedToReject) => {
  const url = `${baseUrl}/push/${id}/reject`;
  let errorMsg = '';
  let isUserAllowedToReject = true;
  await axios.post(url, {}, getAxiosConfig()).catch((error) => {
    if (error.response && error.response.status === 401) {
      errorMsg = 'You are not authorised to reject...';
      isUserAllowedToReject = false;
    }
  });
  await setMessage(errorMsg);
  await setUserAllowedToReject(isUserAllowedToReject);
};
```

This is the related API endpoint (`src/service/routes/push.ts`):

```ts
router.post('/:id/reject', async (req: Request, res: Response) => {
  if (req.user) {
    const id = req.params.id;

    // Get the push request
    const push = await getValidPushOrRespond(id, res);
    if (!push) return;

    // Get the committer of the push via their email
    const committerEmail = push.userEmail;
    const list = await db.getUsers({ email: committerEmail });

    if (list.length === 0) {
      res.status(401).send({
        message: `There was no registered user with the committer's email address: ${committerEmail}`,
      });
      return;
    }

    if (
      list[0].username.toLowerCase() === (req.user as any).username.toLowerCase() &&
      !list[0].admin
    ) {
      res.status(401).send({
        message: `Cannot reject your own changes`,
      });
      return;
    }

    const isAllowed = await db.canUserApproveRejectPush(id, (req.user as any).username);
    console.log({ isAllowed });

    if (isAllowed) {
      const result = await db.reject(id, null);
      console.log(`user ${(req.user as any).username} rejected push request for ${id}`);
      res.send(result);
    } else {
      res.status(401).send({
        message: 'User is not authorised to reject changes',
      });
    }
  } else {
    res.status(401).send({
      message: 'not logged in',
    });
  }
});
```

Notice how all errors must be set to 401 in order to get an error message to display on the frontend. On top of that, the actual reason is not passed to the frontend.

**Describe the solution you'd like**
We could scope this issue down to these core tasks:
- Fixing up the UI error display so that **any** API error is correctly displayed (popup/toast)
- Fixing up the API call functions in the UI so that errors are caught and their messages are displayed regardless of error code

This might require some deeper refactoring so that we don't rely on callback functions for setting the error messages on the UI. We should probably use some kind of reactive, standalone error components.

### LINKED ISSUE #1392: Poor error messaging in the admin UI when the user is not logged in
**Describe the bug**
The admin UI has poor error messaging when the user is not logged in, in some cases. For example, a reviewer attempting to approve or reject a push is told 'You are not authorized to approve/reject" rather than being told they are not logged in or redirected to the login form.

Similarly an a committer attempting to cancel a commit is told 'You are not authorized to cancel" rather than being told they are not logged in or redirected to the login form.

**To Reproduce**

1. Attempt to reject, approve or cancel a push when not logged in.

**Expected behavior**
An accurate error message is reports OR (better) the user is redirected to the login form.

### PR DESCRIPTION
**Pull request by dcoric**
## Summary

This PR fixes both:
- Fixes #1392
- Fixes #1188

## Improve error handling and test coverage for UI services

This PR refactors error handling across the UI service layer and adds comprehensive test coverage. Error handling is now consistent, type-safe, and properly tested.

### What's changed

**Error handling improvements:**
- Added `src/ui/services/errors.ts` with reusable error utilities (`getServiceError`, `errorResult`, `successResult`)
- Refactored `git-push.ts` and `repo.ts` services to use the new error handling utilities
- Updated UI views (`PushDetails.tsx`, `RepoDetails.tsx`, etc.) to properly handle and display error states
- All service functions now return consistent `ServiceResult<T>` objects with success/failure states

**Test coverage (14 → 79 tests):**
- `test/ui/errors.test.ts` - comprehensive tests for error utility functions (18 tests)
- `test/ui/user.test.ts` - tests for user service functions (13 tests)
- `test/ui/git-push.test.ts` - expanded coverage for all git-push functions (13 tests)
- `test/ui/repo.test.ts` - expanded coverage for all repo functions (21 tests)

Tests cover success paths, HTTP errors (401/403/404/409/500), network failures, and edge cases.

### Benefits

- **Consistent error handling** - all services use the same pattern
- **Better UX** - proper error messages shown to users
- **Type safety** - TypeScript ensures correct error handling
- **Test coverage** - prevents regressions and documents expected behavior

Screenshot 2026-02-13 at 18 20 59

## Addition to this PR - Error Boundary

New file `src/ui/components/ErrorBoundary/ErrorBoundary.tsx`

ErrorBoundary class component with getDerivedStateFromError (render capture) + componentDidCatch (logging)

Wraps all route content in <ErrorBoundary name='Dashboard'> and catches any render error from any view without taking down the sidebar/navbar 

It has 2 modes - `dev` and `production`. In `dev` mode it has option to display stacktrace of the error, in production mode only offers option to reload:

DEV: 
Screenshot 2026-02-20 at 17 10 46

PROD:
Screenshot 2026-02-20 at 17 10 58

### CODE REVIEW

File: src/ui/services/git-push.ts [hunk 1]
```diff
@@ -1,133 +1,89 @@
...
+import { getApiV1BaseUrl } from './apiConfig';
 import { Action, Step } from '../../proxy/actions';
 import { PushActionView } from '../types';
+import { ServiceResult, errorResult, successResult } from './errors';
 
-const getPush = async (
-  id: string,
-  setIsLoading: (isLoading: boolean) => void,
-  setPush: (push: PushActionView) => void,
-  setAuth: (auth: boolean) => void,
```
Comment (jescalada) [inline src/ui/services/git-push.ts:11]: Was `setAuth` actually doing anything here before? I think it might've been used for handling expired sessions, but it's hard to tell since not all of the old fetch handlers make use of the hook... 🤔

File: src/ui/services/git-push.ts [hunk 2]
```diff
@@ -1,133 +1,89 @@
...
   const apiV1Base = await getApiV1BaseUrl();
   const url = new URL(`${apiV1Base}/push`);
   url.search = new URLSearchParams(query as any).toString();
 
-  setIsLoading(true);
-
   try {
     const response = await axios<Action[]>(url.toString(), getAxiosConfig());
-    setPushes(response.data as PushActionView[]);
+    return successResult(response.data as PushActionView[]);
```
Comment (jescalada) [inline src/ui/services/git-push.ts]: Wonder if there's a way to prevent casting here... It seems that the enhanced `Action` object (`PushActionView`) isn't accurately typed to begin with, since `PushActionView` should be more of a "data container" rather than an actual `Action` with various methods (`addStep`, etc.). That's why we get this error when trying to remove the casting:

```
Type '{ diff: Step; id: string; type: string; method: string; timestamp: number; project: string; repoName: string; url: string; repo: string; steps: Step[]; error: boolean; errorMessage?: string | null | undefined; ... 19 more ...; newIdxFiles?: string[] | undefined; }' is missing the following properties from type 'PushActionView': addStep, getLastStep, setCommit, setBranch, and 5 more.ts(2740)
```

Comment (jescalada) [reply src/ui/services/git-push.ts]: I'd suggest to change the `PushActionView` type to contain the `Action` type but without the unneeded methods - then we could just unpack the `response.data` and initialize a proper `PushActionView` object:

```ts
const getPush = async (id: string): Promise<ServiceResult<PushActionView>> => {
  const apiV1Base = await getApiV1BaseUrl();
  const url = `${apiV1Base}/push/${id}`;

  try {
    const response = await axios<Action>(url, getAxiosConfig());
    const data: Action = response.data;
    const actionView: PushActionView = {
      ...data,
      diff: data.steps.find((x: Step) => x.stepName === 'diff')!,
    };
    return successResult(actionView);
  } catch (error: any) {
    return errorResult(error, 'Failed to load push');
  }
};
```

File: src/ui/services/user.ts [hunk 3]
```diff
@@ -27,14 +28,12 @@ const getUser = async (
     setUser?.(user);
     setIsLoading?.(false);
   } catch (error) {
-    const axiosError = error as AxiosError;
-    const status = axiosError.response?.status;
+    const { status, message } = getServiceError(error, 'Unknown error');
     if (status === 401) {
       setAuth?.(false);
```
Comment (jescalada) [inline src/ui/services/user.ts:33]: Wondering about why we keep `setAuth` here in `user.ts` but not in the other service files 🤔 Might want to double-check that we weren't using `setAuth` for expired session handling or something else entirely.

File: src/ui/services/git-push.ts (hunk 1 above)
Comment (kriswest) [reply src/ui/services/git-push.ts:11]: I think it may have been vestigial... Where it was used in PushDetails.tsx the state variable wasn't actually used:

`const [, setAuth] = useState(true);`

File: src/ui/views/PushDetails/PushDetails.tsx [hunk 4]
```diff
@@ -19,65 +19,85 @@ import TableHead from '@material-ui/core/TableHead';
...
-    }
+    if (!id) return;
+    const load = async () => {
+      setIsLoading(true);
+      const result = await getPush(id);
+      if (result.success && result.data) {
+        setPush(result.data);
+      } else if (result.status === 401) {
+        setIsLoading(false);
+        navigate('/login', { replace: true });
```
Comment (kriswest) [inline src/ui/views/PushDetails/PushDetails.tsx:57]: 👍

File: src/ui/views/PushDetails/PushDetails.tsx [hunk 5]
```diff
@@ -19,65 +19,85 @@ import TableHead from '@material-ui/core/TableHead';
...
-
-  const setUserAllowedToApprove = (userAllowedToApprove: boolean) => {
-    isUserAllowedToApprove = userAllowedToApprove;
-  };
-
-  const setUserAllowedToReject = (userAllowedToReject: boolean) => {
-    isUserAllowedToReject = userAllowedToReject;
+  const handleActionFailure = (result: ServiceResult) => {
+    if (result.status === 401) {
+      navigate('/login', { replace: true });
```
Comment (kriswest) [inline src/ui/views/PushDetails/PushDetails.tsx:42]: 👍

File: src/ui/services/errors.ts [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+  fallbackMessage: string,
+): { status?: number; message: string } => {
+  const status = error?.response?.status;
+  const responseMessage = error?.response?.data?.message;
+  const message =
+    typeof responseMessage === 'string' && responseMessage.trim().length > 0
+      ? responseMessage
+      : error?.message || fallbackMessage;
+  return { status, message };
+};
```
Comment (kriswest) [inline src/ui/services/errors.ts:21]: Happy to see the API determine the message returned to the user. The error messages from the API are generally informative (with a few small exceptions).

I wonder if all you need is a default 'unknown error occurred, response code: XXX' message if the API doesn't return a message. That should mostly happen on 500/server errors I imagine.

File: src/ui/services/git-push.ts (hunk 1 above)
Comment (dcoric) [reply src/ui/services/git-push.ts:11]: Fixed — it was vestigial. As @kriswest noted, the state variable was declared as const [, setAuth] = useState(true) in PushDetails.tsx and the setter was never actually consumed.

File: src/ui/services/git-push.ts (hunk 2 above)
Comment (dcoric) [reply src/ui/services/git-push.ts]: Done - changed `PushActionView` to extends `Omit<Action, ActionMethods>` in `types.ts`, stripping the class methods from the required interface. `getPush` now constructs `actionView` with the spread + explicit `diff` as you suggested, with no casting needed. For `getPushes` (list), the cast is now `as unknown as PushActionView[]` to be explicit that list items don't have `diff` computed - the table view never uses it.

File: src/ui/services/user.ts (hunk 3 above)
Comment (dcoric) [reply src/ui/services/user.ts:33]: The `setAuth` in `user.ts` is intentional, unlike `git-push.ts`, it's actively used to handle 401 responses by setting auth state to `false`, which drives the expired session / redirect-to-login flow. The pattern is different from the vestigial one in `PushDetails`.

File: src/ui/services/errors.ts (hunk 6 above)
Comment (dcoric) [reply src/ui/services/errors.ts:21]: Done. When the API returns a status code but no message, `getServiceError` now returns `Unknown error occurred, response code: ${status}` instead of falling through to the axios error message.

File: src/ui/views/RepoDetails/RepoDetails.tsx [hunk 7]
```diff
@@ -67,22 +82,37 @@ const RepoDetails: React.FC = () => {
...
+    } else if (result.status === 401) {
+      navigate('/login', { replace: true });
+    } else {
+      setIsError(true);
+      setErrorMessage(result.message || 'Failed to refresh repository data');
+    }
   };
 
   const removeRepository = async (id: string) => {
     await deleteRepo(id);
```
Comment (fabiovincenzi) [inline src/ui/views/RepoDetails/RepoDetails.tsx]: Add a try-catch here as well?

File: src/ui/views/RepoDetails/RepoDetails.tsx [hunk 8]
```diff
@@ -67,22 +82,37 @@ const RepoDetails: React.FC = () => {
   const removeUser = async (userToRemove: string, action: 'authorise' | 'push') => {
     if (!repoId) return;
     await deleteUser(userToRemove, repoId, action);
```
Comment (fabiovincenzi) [inline src/ui/views/RepoDetails/RepoDetails.tsx]: We could add a try-catch here to notify the user, since deleteUser can throw, something like:
```
try {
        await deleteUser(userToRemove, repoId, action);
 } catch (err: any) {
        setIsError(true);
        setErrorMessage(err.message || 'Failed to remove user');
        return;
 }
```

**Review by fabiovincenzi — COMMENTED** [review]
Looks good to me, and it covers #1402 , thanks @dcoric!

**Review by jescalada — APPROVED** [review]
